server/src/main/java/org/opensearch/indices/replication/common/ReplicationLuceneIndex.java

```diff
@@ -0,0 +1,523 @@
+            } else {
+                // This flag is used by disk-based allocation to decide whether the remaining bytes measurement is accurate or not; if not
+                // then it falls back on an estimate. There's only a very short window in which the file details are present but incomplete
+                // so this is a reasonable approximation, and the stats reported to the disk-based allocator don't hit this code path
+                // anyway since they always use IndexShard#getRecoveryState which is never transported over the wire.
+                complete = fileMetadataMap.isEmpty() == false;
+            }
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            final FileMetadata[] files = values().toArray(new FileMetadata[0]);
+            out.writeVInt(files.length);
+            for (FileMetadata file : files) {
+                file.writeTo(out);
+            }
+            if (out.getVersion().onOrAfter(StoreStats.RESERVED_BYTES_VERSION)) {
+                out.writeBoolean(complete);
+            }
+        }
+
+        @Override
+        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+            if (params.paramAsBoolean("detailed", false)) {
+                builder.startArray(Fields.DETAILS);
+                for (FileMetadata file : values()) {
+                    file.toXContent(builder, params);
+                }
+                builder.endArray();
+            }
+
+            return builder;
+        }
+
+        public void addFileDetails(String name, long length, boolean reused) {
+            assert complete == false : "addFileDetail for [" + name + "] when file details are already complete";
+            FileMetadata existing = fileMetadataMap.put(name, new FileMetadata(name, length, reused));
+            assert existing == null : "file [" + name + "] is already reported";
+        }
+
+        public void addRecoveredBytesToFile(String name, long bytes) {
+            FileMetadata file = fileMetadataMap.get(name);
+            assert file != null : "file [" + name + "] hasn't been reported";
+            file.addRecoveredBytes(bytes);
+        }
+
+        public FileMetadata get(String name) {
+            return fileMetadataMap.get(name);
+        }
+
+        public void setComplete() {
+            complete = true;
+        }
+
+        public int size() {
+            return fileMetadataMap.size();
+        }
+
+        public boolean isEmpty() {
+            return fileMetadataMap.isEmpty();
+        }
+
+        public void clear() {
+            fileMetadataMap.clear();
+            complete = false;
+        }
+
+        public Collection<FileMetadata> values() {
+            return fileMetadataMap.values();
+        }
+
+        public boolean isComplete() {
+            return complete;
+        }
+    }
+
+    public static class FileMetadata implements ToXContentObject, Writeable {
```

Comment:
Doesn't look like this ever needs to be extended. Let's mark as final.

```suggestion
    public static final class FileMetadata implements ToXContentObject, Writeable {
```